jdaviz/configs/imviz/helper.py

```diff
@@ -105,6 +107,108 @@ def load_data(self, data, parser_reference=None, **kwargs):
+                  error_on_fail=False):
+        """(Re)link loaded data with the desired link type.
+        All existing links will be replaced.
+
+        .. warning::
+
+            Any markers added would be removed. You can add back the markers
+            manually using :meth:`add_markers`. During the markers removal,
+            pan/zoom will also reset.
+
+        Parameters
+        ----------
+        link_type : {'pixels', 'wcs'}
+            Choose to link by pixels or WCS.
+
+        wcs_fallback_scheme : {None, 'pixels'}
+            If WCS linking failed, choose to fall back to linking by pixels or not at all.
+            This is only used when ``link_type='wcs'``.
+            Choosing `None` may result in some Imviz functionality not working properly.
+
+        wcs_use_affine : bool
+            Use affine transform to approximate the WCS alignment, ignoring distortion.
+            This is only used when ``link_type='wcs'``.
+            This is much more performant at the cost of accuracy.
+
+        error_on_fail : bool
+            If `True`, any failure in linking will raise an exception.
+            If `False`, warnings will be emitted as snackbar messages.
+            When only warnings are emitted and no links are assigned,
+            some Imviz functionality may not work properly.
+
+        Raises
+        ------
+        ValueError
+            Invalid inputs or reference data.
+
+        """
+        if len(self.app.data_collection) <= 1:  # No need to link, we are done.
+            return
+
+        if link_type not in ('pixels', 'wcs'):
+            raise ValueError(f"link_type must be 'pixels' or 'wcs', got {link_type}")
+        if link_type == 'wcs' and wcs_fallback_scheme not in (None, 'pixels'):
+            raise ValueError("wcs_fallback_scheme must be None or 'pixels', "
+                             f"got {wcs_fallback_scheme}")
+
+        # Clear any existing markers. Otherwise, re-linking will crash.
+        self.reset_markers()
+
+        refdata, iref = get_reference_image_data(self.app)
+        links_list = []
+        ids0 = refdata.pixel_component_ids
+        ndim_range = range(refdata.ndim)
+
+        for i, data in enumerate(self.app.data_collection):
+            # Do not link with self
+            if i == iref:
+                continue
+
+            # We are not touching any existing Subsets. They keep their own links.
+            if not layer_is_image_data(data):
+                continue
+
+            ids1 = data.pixel_component_ids
+            try:
+                if link_type == 'pixels':
+                    new_links = [LinkSame(ids0[i], ids1[i]) for i in ndim_range]
+                else:  # 'wcs'
+                    wcslink = WCSLink(data1=refdata, data2=data, cids1=ids0, cids2=ids1)
+                    if wcs_use_affine:
+                        new_links = [wcslink.as_affine_link()]
```

Comment:
If this fails with a NoAffineApproximation error (see link guide) I would fall back on the full solution. Then in the docstring just make it clear that if a good enough approximation can't be found it will use the full WCS transformation.

Reply:
Okay. I added a logic for that. Please double check.

Reply:
Looks good now!